src/tree.c

```diff
@@ -361,6 +361,26 @@ static int parse_mode(uint16_t *mode_out, const char *buffer, size_t buffer_len,
 	int32_t mode;
 	int error;
 
+	if (buffer_len >= 7) {
```

Comment:
I think this is to speed up index reading. 6% performance increase if I recall correctly.

Basically, this function has an expensive implementation but it virtually always gets called with just 3 different arguments, so I special cased them.